test/segment-loader.test.js

```diff
@@ -782,6 +809,82 @@ QUnit.test('tracks segment end times as they are buffered', function(assert) {
   assert.equal(loader.mediaRequests, 1, '1 request');
 });
 
+QUnit.test('adds cues with segment information to the segment-metadata track as they are buffered',
+  function(assert) {
+    const track = loader.segmentMetadataTrack_;
+    let playlist = playlistWithDuration(40);
+    let probeResponse;
+    let expectedCue;
+
+    loader.addSegmentMetadataCue_ = ogAddSegmentMetadataCue_;
+    loader.syncController_.probeTsSegment_ = function(segmentInfo) {
+      return probeResponse;
+    };
+
+    loader.playlist(playlist);
+    loader.mimeType(this.mimeType);
+    loader.load();
+    this.clock.tick(1);
+
+    assert.ok(!track.cues.length, 'segment-metadata track empty when no segments appended');
+
+    // Start appending some segments
+    probeResponse = { start: 0, end: 9.5 };
+    this.requests[0].response = new Uint8Array(10).buffer;
+    this.requests.shift().respond(200, null, '');
+    mediaSource.sourceBuffers[0].trigger('updateend');
+    this.clock.tick(1);
+    expectedCue = {
+      uri: '0.ts',
+      timeline: 0,
+      playlist: 'playlist.m3u8',
+      start: 0,
+      end: 9.5
+    };
+
+    assert.equal(track.cues.length, 1, 'one cue added for segment');
+    assert.deepEqual(JSON.parse(track.cues[0].value.data), expectedCue,
+      'added correct segment info to cue');
+
+    probeResponse = { start: 9.56, end: 19.2 };
+    this.requests[0].response = new Uint8Array(10).buffer;
+    this.requests.shift().respond(200, null, '');
+    mediaSource.sourceBuffers[0].trigger('updateend');
+    this.clock.tick(1);
+    expectedCue = {
+      uri: '1.ts',
+      timeline: 0,
+      playlist: 'playlist.m3u8',
+      start: 9.56,
+      end: 19.2
+    };
+
+    assert.equal(track.cues.length, 2, 'one cue added for segment');
+    assert.deepEqual(JSON.parse(track.cues[1].value.data), expectedCue,
+      'added correct segment info to cue');
+
+    probeResponse = { start: 19.24, end: 28.99 };
+    this.requests[0].response = new Uint8Array(10).buffer;
+    this.requests.shift().respond(200, null, '');
+    mediaSource.sourceBuffers[0].trigger('updateend');
+    this.clock.tick(1);
+    expectedCue = {
+      uri: '2.ts',
+      timeline: 0,
+      playlist: 'playlist.m3u8',
+      start: 19.24,
+      end: 28.99
+    };
+
+    assert.equal(track.cues.length, 3, 'one cue added for segment');
+    assert.deepEqual(JSON.parse(track.cues[2].value.data), expectedCue,
+      'added correct segment info to cue');
+
+    // verify stats
+    assert.equal(loader.mediaBytesTransferred, 30, '10 bytes');
+    assert.equal(loader.mediaRequests, 3, '1 request');
+  });
```

Comment:
Adding a test for overlapping segments is definitely a good call. I don't think being on a different timeline makes a different and needs to be tested. At the point of creating the cues, the start and end times have been converted to display time, so the non-congruent timestamps across discontinuities are not a factor